powershell_kernel/powershell_proxy.py

```diff
@@ -25,44 +25,46 @@ def run(self):
 
 class ReplProxy(object):
     def __init__(self, repl):
+        self.runCmdLock = Lock()
+
         self._repl = repl
         self._repl_reader = ReplReader(repl)
-        # this is a hack to detect when we stop processing this input
-        self.send_input('function prompt() {"^"}')
 
         self.stop_flag = False
         self.output = ''
         self.timer = Timer(0.1, self.update_view_loop)
         self.timer.start()
-        # get preambula and eveluation of the prompt
-        self.get_output()
 
         self.output_prefix_stripped = True
         self.expected_output_prefix = ''
         self.expected_output_len = 0
 
+        # this is a hack to detect when we stop processing this input
+        self.run_command('function prompt() {"^"}')
 
-    def get_output(self):
-        while not self.stop_flag:
-            sleep(0.05)
-        out = self.output
-        self.output = ''
-        self.stop_flag = False 
-        return out       
-
-    def send_input(self, input):
-        # TODO: we should block here until we return output for previous command, should we?
-
-        # for multiline statements we should send 1 extra new line
-        # https://stackoverflow.com/questions/13229066/how-to-end-a-multi-line-command-in-powershell
-        if '\n' in input:
-            input += '\n'
-
-        self.expected_output_prefix = input.replace('\n', '\n>> ') + '\n'
-        self.expected_output_len = len(self.expected_output_prefix)
-        self.output_prefix_stripped = False
-
-        self._repl.write(input + '\n')
+    def run_command(self, input):
+        self.runCmdLock.acquire()
+        try:
+            self.output = ''
+            self.stop_flag = False
+
+            # Append newline to the original input to handle single line comments on the last line
+            #
+            # Also, for multiline statements we should send 1 extra new line at the end
+            # https://stackoverflow.com/questions/13229066/how-to-end-a-multi-line-command-in-powershell
+            input = '. {\n' + input + '\n}\n'
```

Comment:
I noticed a few strings in this method aren't created with string formatting. It's totally optional but I find it helps make strings a little more readable.

Reply:
In this case there are some curly braces in the original string, so using format() makes the string even more confusing. I'll leave this one as is.